File: src/sql/project.py
#!/usr/bin python3
# -*- coding: utf-8 -*-
"""
    :copyright: Copyright 2022 by the FreePDM team
    :license:   MIT License.
"""

from sqlalchemy.orm import sessionmaker
from database import Base
import typing


# Basically searching in the SQL database requires a coonetcion to the database too.
# Is it than a sub-class or better to reqest acces to a class or has this class a interface too?
class Project():
    """Search class"""
    # https://docs.sqlalchemy.org/en/14/orm/session_basics.html

    def __init__(self):
        print("Search in Databases")

    def search_number(self, number: int):
        """Search on number"""
        raise NotImplementedError("Function search_number is not implemented yet")

    def search_description(self, description: str):
        """Search on description"""
        raise NotImplementedError("Function search_description is not implemented yet")

    def search_something_else(self, something: str):
        """Search on something else"""
        raise NotImplementedError("Function search_something_else is not implemented yet")

    def search_help(self) -> str:
        """Print help function"""
        help_text = """
        ad some text
        - Modifiers
        - Search keys etc
        """
        print(help_text)

File: pkg/db/project.go
// Copyright 2023 The FreePDM team. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

package db

import "gorm.io/gorm"

// Basically searching in the SQL database requires a connection to the database too.
// Is it than a sub-class or better to request access to a class or has this class a interface too?

// Project related struct
type Project struct {
	gorm.Model
	number int
	name   string
	status ProjectState
	path   string
}

// Get id with Project number
func (self Project) GetId(number string) int {
	// raise NotImplementedError("Function get_id is not implemented yet")

	return -1 // BUG: Watch out !!!
}

var number int

// Create new Project number
// number is only added when it has a value.
func createNumber() int {

	// increase the number
	number += 1

	return number
}

// Create new project
func NewProject(number *int, name string, status *string, path string) (*Project, error) {
	proj := new(Project)
	proj.name = name
	proj.path = path // TODO: create path automatically
	// TODO: How to handle other related properties

	if number == nil {
		// TODO: get latest number and ndigits from conf / db
		proj.number = createNumber()
	} else {
		proj.number = *number
	}

	if status == nil {
		proj.status = New
	}

	// new_project = PdmProject(self.number, self.name, self.status, self.date_start, self.date_finish, self.path)

	// // TODO: Import Engine - From where?
	// Session.configure(bind=engine, future=True)

	// // https://docs.sqlalchemy.org/en/14/orm/session_basics.html#id1
	// with Session() as session:
	//     try:
	//         session.add(new_project)
	//     except:
	//         Session.rollback()
	//     finally:
	//         Session.close()

	return proj, nil
}

// Remove existing project
func (p *Project) RemovePproject() {
	// raise NotImplementedError("Function remove_project is not implemented yet")
}

// Update existing project
func (p *Project) UpdateProject() {
	// raise NotImplementedError("Function update_model is not implemented yet")
}

// Add user to project
func (p *Project) AddUserToProject() {
	// raise NotImplementedError("Function add_user_to_project is not implemented yet")
}

// Remove user from project
func (p *Project) RemoveUserFromProject() {
	// raise NotImplementedError("Function remove_user_from_project is not implemented yet")
}
